libbpf-tools/syscount.c

```diff
@@ -0,0 +1,1167 @@
+		if (err) {
+			warn("failed to delete element: %s\n", strerror(errno));
+			return false;
+		}
+	}
+
+	*count = i;
+	return true;
+}
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int number;
+	int err;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'x':
+		env.failures = true;
+		break;
+	case 'L':
+		env.latency = true;
+		break;
+	case 'm':
+		env.milliseconds = true;
+		break;
+	case 'P':
+		env.process = true;
+		break;
+	case 'p':
+		err = get_int(arg, &env.pid, 1, INT_MAX);
+		if (err) {
+			warn("invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'i':
+		err = get_int(arg, &env.interval, 0, INT_MAX);
+		if (err) {
+			warn("invalid INTERVAL: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'd':
+		err = get_int(arg, &env.duration, 1, INT_MAX);
+		if (err) {
+			warn("invalid DURATION: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'T':
+		err = get_int(arg, &env.top, 1, INT_MAX);
+		if (err) {
+			warn("invalid TOP: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'e':
+		err = get_int(arg, &number, 1, INT_MAX);
+		if (err) {
+			number = errno_by_name(arg);
+			if (number < 0) {
+				warn("invalid errno: %s (bad, or can't "
+				     "parse dynamically; consider using "
+				     "numeric value and/or installing the "
+				     "errno program from moreutils)\n", arg);
+				argp_usage(state);
+			}
+		}
+		env.filter_errno = number;
+		break;
+	case 'l':
+		env.list_syscalls = true;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static volatile sig_atomic_t hang_on = 1;
+
+void sig_int(int signo)
+{
+	hang_on = 0;
+}
+
+int main(int argc, char **argv)
+{
+	int (*compar)(const void *, const void *);
+	void (*print)(struct data_t *, size_t);
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct data_t vals[MAX_ENTRIES];
+	struct syscount_bpf *obj;
+	int seconds = 0;
+	__u32 count;
+	int err;
+
+	init_syscall_names();
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		goto free_names;
+
+	if (env.list_syscalls) {
+		list_syscalls();
+		goto free_names;
+	}
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		warn("failed to increase rlimit: %s\n", strerror(errno));
+		goto free_names;
+	}
+
+	obj = syscount_bpf__open();
+	if (!obj) {
+		warn("failed to open and/or load BPF object\n");
+		err = 1;
+		goto free_names;
+	}
+
+	if (env.pid)
+		obj->rodata->filter_pid = env.pid;
+	if (env.failures)
+		obj->rodata->filter_failed = true;
+	if (env.latency)
+		obj->rodata->measure_latency = true;
+	if (env.process)
+		obj->rodata->count_by_process = true;
+	if (env.filter_errno)
+		obj->rodata->filter_errno = env.filter_errno;
+
+	err = syscount_bpf__load(obj);
+	if (err) {
+		warn("failed to load BPF object: %s\n", strerror(-err));
+		goto cleanup_obj;
+	}
+
+	err = syscount_bpf__attach(obj);
```

Comment:
so here you don't have to attach everything, you can control which programs need to be attached and then use 
```
bpf_program__attach(obj->progs.whichone);
```

Reply:
Right!